library/core/src/sync/exclusive.rs

```diff
@@ -0,0 +1,130 @@
+//! Defines [`Exclusive`].
+
+use core::future::Future;
+use core::pin::Pin;
+use core::task::{Context, Poll};
+
+/// `Exclusive` provides only _mutable_ access, also referred to as _exclusive_
+/// accesss to the underlying value. It provides no _immutable_, or _shared_
+/// access to the underlying value.
+///
+/// While this may seem not very useful, it allows `Exclusive` to _unconditionally_
+/// implement [`Sync`]. Because _nothing_ can be done with an `&Exclusive`, it can
+/// trivially be send between threads, which satifies `Sync`'s safety requirements.
+///
+/// Certain constructs like [`Future`]s can only be used with _exclusive_ access,
+/// and often `Send` but not `Sync`, so `Exclusive` can be used as hint to the
+/// rust compiler that something is `Sync` in practice.
+///
+/// ## Examples
+/// Using a non-`Sync` future prevent the wrapping struct from being `Sync`
+/// ```compile_fail
+/// use core::cell::Cell;
+///
+/// async fn other() {}
+/// fn assert_sync<T: Sync>(t: T) {}
+/// struct State<F> {
+///     future: F
+/// }
+///
+/// assert_sync(State {
+///     future: async {
+///         let cell = Cell::new(1);
+///         let cell_ref = &cell;
+///         other().await;
+///         let value = cell_ref.get();
+///     }
+/// });
+/// ```
+///
+/// `Exclusive` ensure the struct is `Sync` without stripping the future of any
+/// functionality.
+/// ```
+/// #![feature(exclusive_struct)]
+/// use core::cell::Cell;
+/// use core::sync::Exclusive;
+///
+/// async fn other() {}
+/// fn assert_sync<T: Sync>(t: T) {}
+/// struct State<F> {
+///     future: Exclusive<F>
+/// }
+///
+/// assert_sync(State {
+///     future: Exclusive::new(async {
+///         let cell = Cell::new(1);
+///         let cell_ref = &cell;
+///         other().await;
+///         let value = cell_ref.get();
+///     })
+/// });
+/// ```
+///
+/// ## Parallels with a mutex
+/// In some sense, `Exclusive` can be thought of as a _compile-time_ version of
+/// a mutex, as the borrow-checker guarantees that only one `&mut` can exist
+/// for any value. This is a parallel with the fact that
+/// `&` and `&mut` references together can be thought of as a _compile-time_
+/// version of a read-write lock.
+///
+///
+/// [`Sync`]: core::marker::Sync
+#[unstable(feature = "exclusive_struct", issue = "none")]
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+#[repr(transparent)]
```

Comment:
Why is this `#[repr(transparent)]`?

Reply:
Given the design, there is no reason not to be, ever, so if we can provide this guarantee it could be nice for other code. But then I think it should be accompanied by the utility constructor:

```rust
// and-or a `From<&'lt mut T> for &'lt mut Exclusive<T>`
impl<T : ?Sized> Exclusive<T> {
    pub fn from_mut(r: &'_ mut T) -> &'_ mut Exclusive<T> {
        // SAFETY: repr is ≥ C, so refs have the same layout; and `Exclusive` properties are `&mut`-agnostic
        unsafe { &mut *(r as *mut T as *mut Exclusive<T>) }
    }
}
```

Reply:
And a `from_pin_mut` as well, I guess.

Both would be useful for the `[Pin<] &mut dyn Future [+ Unpin]` cases, where there would be no `Sync` _per se_ but we could end up needing it; we skip the layer of indirection that `Exclusive::new()` would have us go through 🙂

Reply:
I think `from_mut` is nicer than adding another from impl, id rather keep the `From<T: ?Sized>` impl for the common case!

Reply:
As long as it is not documented, it is not considered a public guarantee.